### ISSUE TITLE
C3 editor is unusable if .c3addon with bug (aparently?) is imported

### ISSUE THREAD

**Issue by locohost** [Scirra/Construct-3-bugs#972]
## Problem description

*I made one small modification to the example .c3addon file. Add my own "alert" method to plugin prototype in runtime.js, updated aces.json and lang-en.json according to instructions. I zipped the contents of the plugin folder and renamed to .c3addon. In C3 I openned addon manager and imported the plugin. It imported successfully with no errors reported. I closed FF and restarted editor, now both the "addon manager" and the "insert object" dialogs show "Loading...." forever. They never load and I see no error message.*

## Attach a .c3p

*Attached*

## Steps to reproduce

1. Create a simple plugin with some error? Maybe in one of the .json files? No way for me to know.
2. Import the new addon in C3
3. Save the project
4. Close Firefox, reopen C3 editor
5. Open Addon Manager and/or Insert Object
6. Dialogs show "Loading..." forever

## Observed result

*Addon Manager and Insert Object dialogs show "Loading..." forever. Now I cannot correct the error. C3 editor now no longer has an addon manager and no longer can insert objects. It's unusable and I do not know how to correct.*

## Expected result

*The AddOn manager should be able to detect and plugin errors during import and notify info about the error then not allow the import.*

## Affected browsers

* Chrome: doesn't run C3 correctly
* Firefox: yes
* Edge: don't have
* Safari: don't have

## System details

*If you see a crash report dialog, please copy and paste it to where it says "PASTE HERE" below. Otherwise please go to **Menu** -> **About** -> **Platform information** and paste that information there instead.*

<details><summary>View details</summary>

Platform information

Browser: Firefox
Browser version: 56.0
Browser engine: Gecko
Browser architecture: 64-bit
Context: browser
Operating system: Windows
Operating system version: 8.1
Operating system architecture: 64-bit
Device type: desktop
Device pixel ratio: 1.25
Logical CPU cores: 8
Approx. device memory: (unavailable)
User agent: Mozilla/5.0 (Windows NT 6.3; Win64; x64; rv:56.0) Gecko/20100101 Firefox/56.0
C3 release: r64
Language setting: en-US
Local storage

Storage quota (approx): (status unavailable)
Storage usage (approx): (status unavailable)
Persistant storage: (status unavailable)
Browser support notes

This list contains missing features that are not required, but could improve performance or user experience if supported.

    Rendering multiple on-screen Layout Views is slow in Firefox due to bug 1163426
    UI animations are disabled in Firefox due to bug 1354501
    CSS containment is not supported. Editor performance may be significantly degraded.
    The <dialog> element is not supported. A polyfill is in use.
    This browser does not support pluralisation rules. Plural forms may not be correct.
    ImageBitmapOptions is not supported. Texture loading performance may be degraded.
    Determining input device capabilities is not supported.
    Persistent storage is not available. Local storage could be deleted by the browser.
    Storage quota estimate is unavailable.
    image-set() is not supported. Mouse cursors may appear with reduced quality.

WebGL information

Version string: WebGL 2.0
Numeric version: 2
Supports NPOT textures: yes
Supports GPU profiling: yes
Vendor: Google Inc.
Renderer: ANGLE (Intel(R) HD Graphics 4600 Direct3D11 vs_5_0 ps_5_0)
Major performance caveat: no
Maximum texture size: 16384
Point size range: 1 to 1024
Extensions:

    EXT_color_buffer_float
    EXT_texture_filter_anisotropic
    EXT_disjoint_timer_query
    OES_texture_float_linear
    WEBGL_compressed_texture_s3tc
    WEBGL_debug_renderer_info
    WEBGL_debug_shaders
    WEBGL_lose_context
    MOZ_WEBGL_lose_context
    MOZ_WEBGL_compressed_texture_s3tc

Audio information

System sample rate: 44100 Hz
Output channels: 2
Output interpretation: speakers
Supported decode formats:

    Ogg Opus (audio/ogg; codecs=opus)
    WebM Opus (audio/webm; codecs=opus)
    Ogg Vorbis (audio/ogg; codecs=vorbis)
    WebM Vorbis (audio/webm; codecs=vorbis)
    MPEG-4 AAC (audio/mp4; codecs=mp4a.40.5)
    MP3 (audio/mpeg)
    FLAC (audio/flac)
    PCM WAV (audio/wav; codecs=1)

Supported encode formats:

    Ogg Opus (audio/ogg; codecs=opus)

Video information

Supported decode formats:

    WebM VP9 (video/webm; codecs=vp9)
    WebM VP8 (video/webm; codecs=vp8)
    Ogg Theora (video/ogg; codecs=theora)
    H.264 (video/mp4; codecs=avc1.42E01E)

Supported encode formats:

    WebM VP8 (video/webm; codecs=vp8)

</details>
[pic-quiz-puzzle-game.zip](https://github.com/Scirra/Construct-3-bugs/files/1418676/pic-quiz-puzzle-game.zip)

**Comment by locohost** [comment]
Here is the actual .c3addon (zip)

[SetImageSrcFromBase64.zip](https://github.com/Scirra/Construct-3-bugs/files/1422029/SetImageSrcFromBase64.zip)

**Comment by locohost** [comment]
Tried clearing out the FF memory objects. Was hoping that would be a simulation of "uninstall and re-install C3", but I was wrong. Now C3 freezes on splash/loading screen at 99% loading addons. So what should i do?...

**Comment by Scoremonger** [comment]
I don't know about Firefox, but I was in a similar position with Chrome - a plugin I made has something wrong with it and was causing Construct to be unable to open. I was able open C3 again in Chrome by going to Settings->Advanced->Privacy and Security->Clear browsing data->and checking "Cookies and other site data", then hitting the CLEAR BROWSER DATA button. This nukes all your installed plugins and you login info, so you'll have to re-set up everything again, but if it's a plugin causing the failure this should at least let you open Construct again.

It's surprising that Construct doesn't handle this situation more gracefully. Welcome to the bleeding edge, I guess. ;)

**Comment by locohost** [comment]
Thanks Scoremonger! I'll give that a try later today. Hey, what version of Chrome are you using? I have latest 62 version and the Sprite Editor looks terrible, broken unusable display. That's why I have to use FF for C3. Not by choice :-/

**Comment by Scoremonger** [comment]
Look like I'm still on Version 61.0.3163.100 (Official Build) (64-bit). I'll update and see if the sprite editor is broken for me too.

**Comment by Scoremonger** [comment]
OK, I updated to Version 62.0.3202.75 (Official Build) (64-bit), and the sprite editor looks fine with vanilla C3 on my machine. I guess if clearing the browser data in Chrome doesn't fix things, perhaps reinstalling Chrome will? That's my go-to "turn it off and back on again" solution, anyway. ;)

**Comment by locohost** [comment]
I cleared all browser history in Firefox and C3 starts up again. Thanks for the tip Scoremonger! :-) 

I was able to load the skeleton example plugin, reload C3 and the editor still works fine. So one of the veeeery minor changes I made to the skeleton example plugin completely blows up the editor addons import and insert object UIs when you import it then reload C3. 

C3 doesn't have a few features I need, I'm a looooong time JavaScript programmer, C3 has a plugin SDK so I build some plugins to suit my needs right? Well no, not really, not at the moment. How do I learn what is breaking my simple test plugin?

Do I make any assumptions from the fact no one from Scirra has commented in 4 days?

**Comment by Scoremonger** [comment]
You're welcome locohost, I'm glad that worked! I asked a similar question in the SDK forum about debugging - I would have expected an addon so broken that C3 can't start would prompt a failure message to give me some clue about what the problem is, but no such luck. I guess C3 doesn't do much validation of new addons. So far I've gotten no answer on the forums either. 

As for the lack of responses from Scirra on the bugs, it's frustrating but I wouldn't read _too_ much into it. I saw in another bug report that Diego is on vacation, so at least one key engineer is away from their office. Hopefully it's not a months-long sabbatical!

I ended up starting over from scratch with my addon, changing one little bit of the behavior template at a time, loading it to make sure it didn't crash C3, repeating and gradually adding my functionality back in. The workflow is terrible but I'm making progress that way.

**Comment by locohost** [comment]
Yeah that's good advice. I'll do the same and slowly update the example plugin and I'll add one more step. I'll carefully document each change/add to the example that way when it crashes again, at least I can post my steps in the Scirra SDK forum. Maybe someone will read that and see what I forgot to do or did wrong.

**Comment by locohost** [comment]
I think I read somewhere, maybe in the SDK manual, Ashley implemented the "serve your C3 plugin from a local web server" feature. Apparently C3 will find it (only on localhost) a load it on each app run. No need to keep re-importing/re-starting. I'll setup a little Node/Express server and try this later. Maybe tomorrow. Halloween is tonight!

**Comment by locohost** [comment]
I'll post this to the Scirra SDK forum too. I started with the base example plugin and mades these EXACT edits. Then I zipped just the contents of the plugin folder and imported into C3 (after I cleared FF history to reset C3 editor). **After import then restart C3, I get exact same behavior as OP above.**

## - Edits to plugin.js
  - Changed text on line 12: const PLUGIN_ID = "Test1Plugin";
  - Changed name, twice on line 20: const PLUGIN_CLASS = SDK.Plugins.Test1Plugin = class Test1Plugin extends SDK.IPluginBase
  - Changed text on line 26: SDK.Lang.PushContext("plugins.test1plugin");

## - Edits to type.js
  - Changed name on line 4: const PLUGIN_CLASS = SDK.Plugins.Test1Plugin;
  - Changed name on line 6: PLUGIN_CLASS.Type = class Test1PluginType extends SDK.ITypeBase


## - Edits to instance.js
  - Changed name on line 4: const PLUGIN_CLASS = SDK.Plugins.Test1Plugin;
  - Changed name on line 6: PLUGIN_CLASS.Instance = class Test1Instance extends SDK.IInstanceBase

## - Edits to addon.js
  - Changed text on line 5: "id": "Test1Plugin",

## - Edits to runtime.js
  - Added 1 new action "NewAlert" on line 86 just below the existing "Alert" function
```
      Acts.prototype.NewAlert = function (mssg)
      {
          alert("Test1 NewAlert property = " + mssg);
      };
```
## - Edits to aces.json
  - Added 1 new action "do-new-alert" on line 22 just below the existing "do-alert" JSON
  - Commas are correct before, within and after the new JSON
```
      {
          "id": "do-new-alert",
          "scriptName": "NewAlert",
          "highlight": true,
          "params": [
              {
                  "id": "new-mssg",
                  "type": "string"
              }
          ]
      },
```
## - Edits to en-US.json
  - Changed text on line 6: "test1plugin": {
  - Added 1 new action "do-new-alert" on line 38 just below the existing "do-alert" JSON
  - Commas are correct before, within and after the new JSON
```
      "do-new-alert": {
          "list-name": "Do New alert",
	  "display-text": "Do New alert",
	  "description": "Do a dummy New alert.",
	  "params": {
	       "new-mssg": {
	           "name": "Message",
	           "desc": "Message to display in New alert."
	       }
	   }
      }
```

**Comment by AshleyScirra** [comment]
Please don't report mistakes in your addons as bugs - they're bugs in your addons, not bugs in Construct.

Errors are reported to the browser console. It tells you you're missing a language string. You also don't need to clear browser storage to remove it. You can use safe mode, or better still, a local development server so you can just fix the problem and press reload. This is all [documented in the Addon SDK](https://www.construct.net/make-games/manuals/addon-sdk), please check that for more information.

**Comment by locohost** [comment]
Yeah not the reply I was expecting from a paid software service. I DID NOT report a bug in my code. I reported your app UI breaks when anyone uploads a plugin with a bug. **This a bug in your C3 editor code.** Your addon manager needs to import any plugin (bugs or not) and report issues just like all other builders on the earth. **It should not accept a plugin upload and then crash the addon and import object UIs like it currently does.**

Are you ACTUALLY saying we must only import perfect plugin code without a single bug every time? Othewise, we have to accept that the C3 UI will always crash and the solution is to clear the browser everytime forever or is there a fix planned?...

**Comment by locohost** [comment]
I never saw an error in console. I do see a GIANT warning in console to not mess with console. I will try the local dev server option.

**Comment by locohost** [comment]
I'd like the "not a bug" tag removed because there is definitely a bug in your addon import. This is what I was reporting.

**Comment by locohost** [comment]
There is no error reported in console on addon import. There is no error reported in console on C3 close/reload. There is no error report in console when I open addon manager and it freezes on "Loading...". There is no error in console when I click Import Object and it freezes on "Loading...". **When and where is this console error you are referring to Ashley?...**

**Comment by locohost** [comment]
Does this error only show when you used "developer mode"?... If yes, you should state all of these things in the documentation.

**Comment by AshleyScirra** [comment]
You are supposed to use developer mode from the start. It is much easier to develop plugins that way. For that reason we assume that once you import a .c3addon file, it is a finished and tested addon. Even so, if there's an issue you can use safe mode to remove it. This is all in the SDK documentation.

**Comment by locohost** [comment]
I will setup a Node/Express server to try developer mode later today. I will look in the SDK docs for specific info **on how to setup the web server files and structure** (not Node/Express, I'm perfectly good with that) too. Thank you for helping.

**Comment by locohost** [comment]
Setup up a Node/Express Https server. It serves the contents of the addon.json file. I enabled developer mode in C3. I tried to load the addon via the dev load button, plugin does not load and console is blank. Url is good and working. Yes I'm entering same url in C3 addon manager UI. See 2 attached screenshots that show the Url returns correctly and C3 screen showing addon load fail and blank console to left/behind.

![c3 addon dev test - firefox https shows addonjson correctly](https://user-images.githubusercontent.com/132458/32273906-c896e2fe-beda-11e7-8d3e-0dd58976eabe.PNG)
![c3 addon dev test - unable to load - blank console](https://user-images.githubusercontent.com/132458/32273915-d370a7b4-beda-11e7-9ea8-090a1b52993a.PNG)
